node/lib/util/checkout.js

```diff
@@ -151,18 +172,306 @@ exports.checkout = co.wrap(function *(metaRepo, committish) {
+                        },
+                    };
+                    result.switchBranch = committish;
+                }
+            }
+
+            // If we didn't resolve anything from `committish`, throw an error.
+
+            if (null === result.commit) {
+                throw new UserError(
+                    `Could not resolve ${colors.red(committish)}.`);
+            }
+        }
+        else {
+
+            const commit = yield repo.getCommit(annotated.id());
+            result.commit = commit;
+
+            // Check to see if the commit refers to a branch name.
+
+            const branch = yield GitUtil.findBranch(repo, committish);
+            if (null !== branch) {
+                committishBranch = branch;
+                if(!branch.isRemote()) {
+                    result.switchBranch = committish;
+                }
+            }
+        }
+    }
+    else {
+        // If we're implicitly using HEAD, see if it's on a branch and record
+        // that branch's name.
+
+        const head = yield repo.head();
+        if (head.isBranch()) {
+            committishBranch = head;
+        }
+    }
+
+    if (null !== newBranch) {
+        // If we have a `newBranch`, we need to make sure it doesn't already
+        // exist.
+
+        yield ensureBranchDoesntExist(newBranch);
+
+        // Now, if we're supposed to set up tracking, validate that the
+        // committish is a branch.  If it's a tracking branch, then parse the
+        // parts.
+
+        let tracking = null;
+        if (track) {
+            // Set up tracking information of `track` is set.
+
+            if (null === committishBranch) {
+                throw new UserError(`\
+Cannot setup tracking information; starting point is not a branch.`);
+            }
+
+            // If the branch is remote, set up remote tracking information,
+            // otherwise leve the remote name 'null';
+
+            if (committishBranch.isRemote()) {
+                const parts = committishBranch.shorthand().split("/");
+                tracking = {
+                    remoteName: parts[0],
+                    branchName: parts[1],
+                };
+            }
+            else {
+                tracking = {
+                    remoteName: null,
+                    branchName: committishBranch.shorthand(),
+                };
+            }
+        }
+
+        result.newBranch = {
+            name: newBranch,
+            tracking: tracking,
+        };
+        result.switchBranch = newBranch;
+    }
+    return result;
+});
+
+/**
+ * Check out the specified `commit`, if provided, in the specified `repo`.
+ * then, if the specified `newBranch` is provided, create a new branch having
+ * the specified `newBranch.name` name.  If the specified `newBranch.tracking`
+ * is provided, configure the newly created branch to track the one it
+ * describes.  If the specified `commit` and `commitBranch` are non-null, but
```

Comment:
This text is pretty confusing.  Can you maybe add a note which explains how it correspond to 'git checkout'?